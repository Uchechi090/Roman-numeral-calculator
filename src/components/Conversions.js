export const convertToRoman = num => {
  var decimalValue = [1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1];
  var romanNumeral = [
    "M",
    "CM",
    "D",
    "CD",
    "C",
    "XC",
    "L",
    "XL",
    "X",
    "IX",
    "V",
    "IV",
    "I"
  ];

  var romanized = "";

  for (var index = 0; index < decimalValue.length; index++) {
    while (decimalValue[index] <= num) {
      romanized += romanNumeral[index];
      num -= decimalValue[index];
    }
  }

  return romanized;
};

const CharToInt = c => {
  switch (c) {
    case "I":
      return 1;
    case "V":
      return 5;
    case "X":
      return 10;
    case "L":
      return 50;
    case "C":
      return 100;
    case "D":
      return 500;
    case "M":
      return 1000;
    default:
      return -1;
  }
};

export const RomanToInt = str1 => {
  if (str1 == null) return -1;
  var num = CharToInt(str1.charAt(0));
  var pre, curr;

  for (var i = 1; i < str1.length; i++) {
    curr = CharToInt(str1.charAt(i));
    pre = CharToInt(str1.charAt(i - 1));
    if (curr <= pre) {
      num += curr;
    } else {
      num = num - pre * 2 + curr;
    }
  }

  return num;
};
