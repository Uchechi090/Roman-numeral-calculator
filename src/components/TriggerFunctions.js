/* eslint-disable no-unused-vars */
/* eslint-disable no-eval */
/* eslint-disable eqeqeq */

export const SyntaxErr = value => {
  if (
    eval(value) == SyntaxError ||
    eval(value) == ReferenceError ||
    eval(value) == TypeError
  ) {
    return value == "Syntax Error";
  }
};

// export const Equals = value => {
//   if (value.indexOf("^") > -1) {
//     var base = value.slice(0, value.indexOf("^"));
//     var exponent = value.slice(value.indexOf("^") + 1);
//     value = eval("Math.pow(" + base + "," + exponent + ")");
//   } else {
//     value = eval(value);
//     //checkLength();
//     SyntaxErr();
//   }
// };

export const Multiply = (x, y) => x * y;

export const Add = (x, y) => x + y;

export const Subtract = (x, y) => x - y;

