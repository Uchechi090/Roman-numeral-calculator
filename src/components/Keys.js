import React, { Component } from "react";
import { RomanToInt, convertToRoman, Compute } from "../components/Conversions";

class Keys extends Component {
  constructor(props) {
    super(props);
    this.state = {
      prevScreen: "",
      prevValue: "",
      currOp: "",
      calValue: "",
      isNum: false,
      isOperand: false
    };
  }

  componentDidMount() {
    this.addEvents();
  }

  addEvents = () => {
    let buttons = document.getElementsByClassName("button");
    Array.prototype.forEach.call(buttons, button => {
      button.addEventListener("click", () => button);
    });
    return buttons;
  };

  assignStateValues = (input, displayValue) => {
    let newValue = "";
    let newInput = "";
    const {
      currOp,
      prevValue,
      prevScreen,
      calValue
    } = this.state;

    
    if (prevScreen === "") {
      newInput = RomanToInt(displayValue);
    } else {
      newInput = RomanToInt(displayValue.substring(prevScreen.length));
    }
    if (prevValue !== "" && calValue === "") {
      newValue = Compute(prevValue, currOp, newInput);
    } else if (calValue !== "") {
      newValue = Compute(calValue, currOp, newInput);
    }
    this.setState({
      prevValue: newInput,
      currOp: input,
      calValue: newValue,
      prevScreen: displayValue + input
    });
  };

  Clear = () => {
    this.setState({
      isNum: false,
      isOperand: false,
      prevScreen: "",
      prevValue: "",
      currOp: "",
      prevOp: "",
      calValue: ""
    });
    return "";
  };

  Equals = displayValue => {
    const {
      calValue,
      currOp,
      currValue,
      prevValue,
      prevScreen
    } = this.state;
    let result = "";
    let newInput = RomanToInt(displayValue.substring(prevScreen.length));
    if (calValue !== "") {
      result = Compute(calValue, currOp, newInput);
    } else if (prevValue !== "") {
      result = Compute(prevValue, currOp, newInput);
    } else if (currValue !== "") {
      return result;
    }

    return convertToRoman(result);
  };

  validateInput = input => {
    const operators = ["+", "-", "*"];
    if (
      operators.includes(input) &&
      !this.state.isNum &&
      !this.state.isOperand
    ) {
      return false;
    }
    if (
      operators.includes(input) &&
      !this.state.isNum &&
      this.state.isOperand
    ) {
      let val = document.getElementById("screen").value;
      val = val.substr(0, val.length - 1) + input;
      document.getElementById("screen").value = val;
      this.setState({ currOp: input });
      return false;
    }
    if (operators.includes(input)) {
      this.setState({ isNum: false, isOperand: true });
      return true;
    }

    if (!operators.includes(input)) {
      if (input === "=" && this.state.isOperand) {
        return false;
      }
      this.setState({ isNum: true, isOperand: false });
      return true;
    }
  };

  handleButtons = e => {
    e.preventDefault();

    let display = document.getElementById("screen");
    let resultScreen = document.getElementById("result");
    const text = e.target.innerText;

    if (this.validateInput(text)) {
      if (text === "Clear") {
        display.value = this.Clear();
        resultScreen.value = this.Clear();
      } else if (text === "=") {
        resultScreen.value = this.Equals(display.value);
      } else if (text === "+") {
        this.assignStateValues(text, display.value);
        display.value += "+";
      } else if (text === "-") {
        this.assignStateValues(text, display.value);
        display.value += "-";
      } else if (text === "*") {
        this.assignStateValues(text, display.value);
        display.value += "*";
      } else {
        display.value += text;
      }
    }
  };

  render() {
    return (
      <div className="calculator" onClick={e => this.handleButtons(e)}>
        <div>
          <input type="text" id="screen" maxLength="20" />
          <input type="text" id="result" maxLength="20" />
        </div>
        <div className="calc-buttons">
          <div className="functions-one">
            <button className="button triggers">Clear</button>
            <button className="button basic-stuff">*</button>
            <button className="button basic-stuff">-</button>
            <button className="button numbers">VII</button>
            <button className="button numbers">VIII</button>
            <button className="button numbers">IX</button>
            <button className="button numbers">IV</button>
            <button className="button numbers">V</button>
            <button className="button numbers">VI</button>
            <button className="button numbers">I</button>
            <button className="button numbers">II</button>
            <button className="button numbers">III</button>
            <button className="button numbers">X</button>
            <button className="button numbers">L</button>
            <button className="button numbers">C</button>
            <button className="button numbers">D</button>
            <button className="button numbers">M</button>
            <button className="button basic-stuff">+</button>
            <button className="button basic-stuff">=</button>
          </div>
        </div>
      </div>
    );
  }
}

export default Keys;
