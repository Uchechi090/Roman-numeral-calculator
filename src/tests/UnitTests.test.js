import React from "react";
import { create } from "react-test-renderer";
import renderer from "react-test-renderer";

import Keys from "../components/Keys";
import { Add, Subtract, Multiply } from "../components/TriggerFunctions";

describe("Keys class component", () => {
  test("it should make a snapshot test", () => {
    const keys = create(<Keys />);
    expect(keys.toJSON()).toMatchSnapshot();
  });
});

test("adds the args together", () => {
  expect(Add(1, 2)).toBe(3);
});

test("subtracts the args together", () => {
  expect(Subtract(4, 2)).toBe(2);
});

test("multiply the args together", () => {
  expect(Multiply(3, 2)).toBe(6);
});
