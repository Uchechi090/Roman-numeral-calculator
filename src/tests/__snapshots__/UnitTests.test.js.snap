// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Keys class component it should make a snapshot test 1`] = `
<div
  className="calculator"
  onClick={[Function]}
>
  <div>
    <input
      id="screen"
      maxLength="20"
      type="text"
    />
    <input
      id="result"
      maxLength="20"
      type="text"
    />
  </div>
  <div
    className="calc-buttons"
  >
    <div
      className="functions-one"
    >
      <button
        className="button triggers"
      >
        Clear
      </button>
      <button
        className="button basic-stuff"
      >
        *
      </button>
      <button
        className="button basic-stuff"
      >
        -
      </button>
      <button
        className="button numbers"
      >
        VII
      </button>
      <button
        className="button numbers"
      >
        VIII
      </button>
      <button
        className="button numbers"
      >
        IX
      </button>
      <button
        className="button numbers"
      >
        IV
      </button>
      <button
        className="button numbers"
      >
        V
      </button>
      <button
        className="button numbers"
      >
        VI
      </button>
      <button
        className="button numbers"
      >
        I
      </button>
      <button
        className="button numbers"
      >
        II
      </button>
      <button
        className="button numbers"
      >
        III
      </button>
      <button
        className="button numbers"
      >
        X
      </button>
      <button
        className="button numbers"
      >
        L
      </button>
      <button
        className="button numbers"
      >
        C
      </button>
      <button
        className="button numbers"
      >
        D
      </button>
      <button
        className="button numbers"
      >
        M
      </button>
      <button
        className="button basic-stuff"
      >
        +
      </button>
      <button
        className="button basic-stuff"
      >
        =
      </button>
    </div>
  </div>
</div>
`;
